### ISSUE TITLE
Phone auth: can't get PhoneAuthCredential if Instant verification or Auto-retrieval

### ISSUE THREAD

**Issue by vdias38** [dpa99c/cordova-plugin-firebasex#435]
Hi Dave, 

thanks a lot for the job you've done maintaining this repo, the quality of code and your support are amazing, congrats!

I've recently faced an issue trying to make phone authentication works for web and native. My app use AngularFire when possible and firebase-x plugin when required, it's the case to handle phone authentication correctly (Instant verification or Auto-retrieval). Then to resume I call `FirebasePlugin.verifyPhoneNumber` to get credential and call `angularFireAuth.auth.signInWithCredential(credential)` to signin. BUT I can't get credential on Android if Instant verification or Auto-retrieval. I've checked the java code and see that you share an id in place of credential in verifyNumber's response.

**Is it possible to set credential in verifyNumber's response when available OR make public a method to get AuthCredential by id** (`FirebasePlugin.instance.authCredentials.get(jsonCredential.getString("id"))`)?

Below an example of how I've updated the code you suggested for my use case. 

```ts
var number = '+441234567890';
var timeOutDuration = 60;
var fakeVerificationCode = '123456';
var awaitingSms = false;

FirebasePlugin.verifyPhoneNumber(function(credential) {

    if(credential.instantVerification){
        if(awaitingSms){
            awaitingSms = false;
            dismissUserPromptToInputCode();
        }   
        this.angularFireAuth.auth.signInWithCredential(credential); // <---- signInWithCredential FAILED with following msg: First argument "credential" must be a valid credential.
    }else{
        awaitingSms = true;
        promptUserToInputCode()
            .then(function(userEnteredCode){
                awaitingSms = false;                
                credential.code = userEnteredCode;
                this.angularFireAuth.auth.signInWithCredential(credential); // <-- Works fine!
            });
    }
}, function(error) {
    console.error("Failed to verify phone number: " + JSON.stringify(error));
}, number, timeOutDuration, fakeVerificationCode);
```

**Comment by skibaalex** [comment]
@vdias38  same issue did you figure this out?

**Comment by cdpndt** [comment]
I am facing same issue #433

**Comment by dpa99c** [comment]
When instant verification is used, `verifyPhoneNumber` on Android [receives a native `PhoneAuthCredential` object](https://github.com/dpa99c/cordova-plugin-firebasex/blob/master/src/android/FirebasePlugin.java#L1408) (this is a similar object to the `AuthCredential` [returned in the Google Sign In flow](https://github.com/dpa99c/cordova-plugin-firebasex/blob/master/src/android/FirebasePlugin.java#L464)).

The original [cordova-plugin-firebase](https://github.com/arnesson/cordova-plugin-firebase), from which this plugin was forked, [attempts to use internal reflection](https://github.com/arnesson/cordova-plugin-firebase/blob/master/src/android/FirebasePlugin.java#L751) to parse the private fields of this object into strings that can be returned up to the JS layer. However this is a fragile strategy which causes numerous edge case problems (e.g. https://github.com/arnesson/cordova-plugin-firebase/issues/983).

Therefore, this plugin instead passes an `id` reference to that native auth object, which can be referenced by the JS in your app when using `signInWithCredential` and `linkUserWithCredential`.

TL;DR: this plugin wraps the native Firebase SDKs and so is not intended to be used interchangeably with the Firebase JS SDK (such as via AngularFire).
This is because the Firebase team does not intend them to be used interchangeably so the native Firebase SDKs return native objects which cannot be reliably parsed into primitive types that can be passed freely between the JS and native layers in a Cordova hybrid app.

**Comment by vdias38** [comment]
Thanks a lot @dpa99c for you well explained response.
I've submitted a question on stackoverflow [How to mix native and JS firebase SDKs to handle phone number authentication](https://stackoverflow.com/questions/62467738/how-to-mix-native-and-js-firebase-sdks-to-handle-phone-number-authentication). The "cross layer authentication" is a widely used pattern and the only pitfall I know concerns phone auth instant verification on Android, so maybe they could do stg to ease it.

**Comment by spoxies** [comment]
One wants to authenticate (using phonenumber), and also be authenticated on the Javascript SDK as it provides functions that `cordova-plugin-firebasex` (currently) does not.

But I'm not clear on what @dpa99c says: "..., this plugin instead passes an `id` reference to that native auth object, which can be referenced by the JS in your app when using signInWithCredential and linkUserWithCredential."

Does this mean it can be achieved as long as one passes this `id` ref to the Javascript (SDK)? And if so how would that look. I'm not able to figure that one out, I might be mislead.

**Comment by vdias38** [comment]
Hi @spoxies , 
the "id referenced by the JS" does not mean JS SDK but JS layer of cordova plugin. The id reference should be used with `signInWithCredential` and `linkUserWithCredential` methods of `firebasex` plugin. Internally firebasex plugin get native PhoneAuthCredential object, referenced by the id, to handle authentication.

**Comment by spoxies** [comment]
@vdias38 I was afraid so. However achieving the "cross layer auth" that was possible before the reworked version (`cordova-plugin-firebasex@8.0.0`) is the blocking issue for you and me I believe. Meaning being able to authenticate on the JS SDK layer as well (after the Native auth). Did someone manage to create a workaround for this?

Because although @dpa99c Writes in https://github.com/dpa99c/cordova-plugin-firebasex/issues/176#issuecomment-558262782 "_there is no reason to extract the credentials from the native Java credential object and return it to the JS layer._" the PR's I mentioned before are created to make login in to the JS SDK possible for two other sign-in options [Sign in with Apple & Sign in with Google] 

Having a similar (workaround/workflow) for between firebasex and JS SDK for PhoneAuth seems quite essential (at least until firebasex will expose everything and I believe that would be to much to ask as the JS SDK will always be steps ahead).

**Comment by vdias38** [comment]
Hi @spoxies, I'm the author of the PR dpa99c/cordova-plugin-firebasex/pull/633 (Sign in with Google), and yes I think it's 
a common case on hybrid app devs to authenticate on native layer and propagate credentials to sign-in with on JS layer. The doc update you've suggested is perfect, maybe you should convert it to a PR for effortless merge.
Last year I've submitted a message to [firebase-talk group](https://groups.google.com/g/firebase-talk) about this topic, but was not published (waiting approval), posted a [stackoverflow issue](https://stackoverflow.com/questions/62467738/how-to-mix-native-and-js-firebase-sdks-to-handle-phone-number-authentication), but was deleted after 1 year without any response :-(
I understand arguments of @dpa99c and it should be a firebase commitment to provide a bridge over web/native layers, if hybrid apps are important for firebase dev team of course... At the moment I can't see a solution for our use case, or should we need to migrate to flutter ? 🤔

**Comment by spoxies** [comment]
@vdias38 Thanks! For Android I'm considering to downgrade to version >7 for Android. There if `instantVerification: true` was also passed with a `credentals.code` and that could be used to pass to Firebase JS SDK and retrieve the credentials. However at this time only iOS is relevant to me.

I do have my hopes that the current maintainer has an opinion/insight about it based on the PR and issues, that will help us both to adjust course accordingly. Personally I think it would be a waste of life(time) to start migration to flutter (and I don't have funds left). So lets see what @dpa99c has to say and if he does not have any deep grained objections. 

If the time comes to focus more on Android I'd rather sponsor this issue because the alternative (migration) seems way worse/time consuming.

p.s. The PR was already pending https://github.com/dpa99c/cordova-plugin-firebasex/pull/644
